SPEC: Add SQL UDF spec (#14117)
apache/iceberg: format/udf-spec.md

---
title: "SQL UDF Spec"
---
<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

# Iceberg SQL UDF Spec

## Background and Motivation

A SQL user-defined function (UDF or UDTF) is a callable routine that accepts input parameters, executes a function body,

> and returns either:

Reviewer: because here we say `return either:`, maybe the bullets should be some thing like

* a single scalar value for scalar functions (UDFs)
* a table  with zero or multiple rows for table functions (UDTFs)

I don't know if zero row is allowed  or not for table functions during the discussion

* **Scalar functions (UDFs):** return a single value.

> * **Table functions (UDTFs):** return a table with one or more rows and columns.

Reviewer 2: I was wondering if a UDTF in this context is just a UDF that returns List<Struct<>>
?

> Most compute engines (e.g., Spark, Trino) already support UDFs, but in different and incompatible ways. Without a common

Reviewer: nit: change `Most` to `Many`

standard, UDFs cannot be reliably shared across engines or reused in multi-engine environments.

This specification introduces a standardized metadata format for UDFs in Iceberg. 

## Goals

* Define a portable metadata format for both scalar and table SQL UDFs. The metadata is self-contained and can be moved across catalogs. 
* Support function evolution through versioning and rollback.
* Provide consistent semantics for representing UDFs across engines.

## Overview

UDF metadata follows the same design principles as Iceberg table and view metadata: each function is represented by a
**self-contained metadata file**. Metadata captures definitions, parameters, return types, documentation, security,
properties, and engine-specific representations.

* Any modification (new overload, updated representation, changed properties, etc.) creates a new metadata file. The old file is replaced atomically.
* Each metadata file includes recent definition versions, enabling rollbacks without external state.

## Specification

### UDF Metadata

> The UDF metadata file has the following fields:

Reviewer: wondering if it is helpful to add a `Field type` to those metadata tables?

I know most tables in view spec don't have the type column. [one table](https://iceberg.apache.org/view-spec/#sql-representation) does include the type column

Reviewer: sounds good. JSON example is also a reference

| Requirement | Field name                   | Description                                                      |
|-------------|------------------------------|------------------------------------------------------------------|

> | *required*  | `function-uuid`              | A UUID that identifies the function, generated once at creation. |

Reviewer: I see. agree that we can keep it consistent with `table-uuid` and `view-uuid`

| *required*  | `format-version`             | Metadata format version (must be `1`).                           |

> | *required*  | `definition`                 | List of function overloads.                                      |

Reviewer: nit: `definitions`?

| *required*  | `definition-versions`        | List of versioned function definitions.                          |
| *required*  | `current-definition-version` | Identifier of the current definition version.                    |
| *optional*  | `location`                   | The storage location of metadata files.                          | 
| *optional*  | `properties`                 | Arbitrary key-value pairs.                                       | 

> | *optional*  | `secure`                     | Whether it is a secure function. Default: `false`.               |

Author: Good point, will add it

|-------------|-----------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| *required*  | `overload-uuid` | A UUID that identifies this function overload.                                                                                                                                                     |
| *required*  | `parameters`    | Ordered list of function parameter definitions (name, type, optional doc). The order of parameters in this list **must exactly match** the order of arguments provided when invoking the function. |
| *required*  | `return-type`   | Return type (scalar or struct). Example: `"string"` or `"struct<...>"`                                                                                                                             |
| *required*  | `versions`      | List of overload versions.                                                                                                                                                                         |

> | *optional*  | `udtf`          | Whether this overload returns a table, default to `false`. If `true`, the `return-type` must be a `struct` describing the output schema.                                                           | 

Reviewer 3: Reminder: We discussed converting this field to a function field type, which will include scalar, aggregate, and table functions.

Note: The `body` must be valid SQL in the specified dialect; validation is the responsibility of the consuming engine.

### Definition-Version

| Requirement | Field name              | Description                                                                                |
|-------------|-------------------------|--------------------------------------------------------------------------------------------|
| *required*  | `definition-version-id` | Unique identifier of the definition version, a monotonically increasing long. Example: `2` |
| *required*  | `timestamp-ms`          | Timestamp when the definition was created or updated. Example: `1734506000456`             |
| *required*  | `body`                  | List of mapping of overload uuids to their current version ids.                            |

## Function Resolution in Engines
Resolution rule is decided by engines, but engines SHOULD:
1. Prefer exact signature matches over casting or subtyping.
2. Allow numeric widening (e.g., `INT` --> `BIGINT/FLOAT`) by default.
3. Require explicit casts for risky conversions (e.g., `STRING` --> `DATE`).

> ## Appendix A: Example

Reviewer: can add another simple example for UDTF and illustrate the return type struct?

      "versions": [
        {
          "overload-version-id": 1,
          "deterministic": true,
          "representations": [
            {
              "type": "sql",
              "dialect": "trino",
              "body": "x + 2"
            }
          ],
          "timestamp-ms": 1734507000123
        },
        {
          "overload-version-id": 2,
          "deterministic": true,
          "representations": [
            {
              "type": "sql",
              "dialect": "trino",
              "body": "x + 1"
            }
          ],
          "timestamp-ms": 1735507000124
        }
      ]
    },
    {
      "overload-uuid": "7c9f93b1-28b4-4ef5-90f5-70c73cda2222",
      "parameters": [
        { "name": "x", "type": "float", "doc": "Input float" }
      ],
      "return-type": "float",
      "doc": "Add one to the input float",
      "versions": [
        {
          "overload-version-id": 1,
          "deterministic": true,
          "representations": [
            {
              "dialect": "trino",
              "body": "x + 1.0"
            }
          ],
          "timestamp-ms": 1734507001123
        }
      ]
    }
  ],

>   "definition-versions": [

Reviewer: I am wondering if we can just add a `current-version` in the overload definition directly, e.g. above line 149?

---
<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

# Iceberg SQL UDF Spec

## Background and Motivation

A SQL user-defined function (UDF or UDTF) is a callable routine that accepts input parameters and executes a function body.
Depending on the function type, the result can be:

- **Scalar functions (UDFs)** – return a scalar value, which may be a primitive type (e.g., `int`, `string`) or a non-primitive type (e.g., `struct`, `list`).

> - **Table functions (UDTFs)** – return a table, i.e., a table with zero or more rows and columns with a uniform schema.

Reviewer 4: ```suggestion
- **Table functions (UDTFs)** – returns a table with zero or more rows of columns with a uniform schema.
```

Many compute engines (e.g., Spark, Trino) already support UDFs, but in different and incompatible ways. Without a common
standard, UDFs cannot be reliably shared across engines or reused in multi-engine environments.

This specification introduces a standardized metadata format for UDFs in Iceberg. 

## Goals

* Define a portable metadata format for both scalar and table SQL UDFs. The metadata is self-contained and can be moved across catalogs. 
* Support function evolution through versioning and rollback.

> * Provide consistent semantics for representing UDFs across engines.

Author: I think this is more about semantics than representation. The goal is to ensure consistent semantics(e.g., same behaviors) across engines, while allowing dialectic representations for each engine. Representations don’t have to be identical, but they should preserve the same meaning. I'm open to suggestion if wording could be improved.

properties, and engine-specific representations.

* Any modification (new overload, updated representation, changed properties, etc.) creates a new metadata file, and atomically swaps in the new file as the current metadata.
* Each metadata file includes recent definition versions, enabling rollbacks without external state.

## Specification

### UDF Metadata
The UDF metadata file has the following fields:

| Requirement | Field name        | Type                   | Description                                                                                                     |
|-------------|-------------------|------------------------|-----------------------------------------------------------------------------------------------------------------|
| *required*  | `function-uuid`   | `string`               | A UUID that identifies the function, generated once at creation.                                                |
| *required*  | `format-version`  | `int`                  | Metadata format version (must be `1`).                                                                          |

> | *required*  | `definitions`     | `list<overload>`       | List of function [overload](#overload) entities.                                                                |

Reviewer 4: These names are a product of the evolution of this spec and I think we are now in a good place to consider choosing more consistent names. It is strange that "definitions" is actually a list of "overloads". I think we should choose a single term for a function signature and use that here. To me, "definition" inside "definitions" makes the most sense.

| *optional*  | `doc`             | `string`               | Documentation string.                                                                                           |

Notes:
1. When `secure` is `true`,
    - Engines **SHOULD NOT** expose the function definition through any inspection (e.g., `SHOW FUNCTIONS`).
    - Engines **SHOULD** ensure that execution does not leak sensitive information through any channels, such as error messages, logs, or query plans.
   
### Overload

Function overloads allow multiple implementations of the same function name with different signatures. Each overload has
the following fields:

| Requirement | Field name                 | Type                                                   | Description                                                                                   |
|-------------|----------------------------|--------------------------------------------------------|-----------------------------------------------------------------------------------------------|

> | *required*  | `overload-id`              | `long`                                                 | Monotonically increasing identifier of this function overload.                                |

Reviewer 4: Do we need an ID for the overload? Isn't it identified by the function parameters?

> | *required*  | `parameters`               | `list<parameter>`                                      | Ordered list of [function parameters](#parameter). Invocation order **must** match this list. |

Author: Added a new note to clarify that parameter names must not change due to named parameter invocation.

> | *required*  | `parameters`               | `array<parameter>`                            | Ordered list of [function parameters](#parameter). Invocation order **must** match this list. |

Author: I was trying to use the json data type name here, https://www.w3schools.com/js/js_json_datatypes.asp.  I'm OK if we agreed to use Iceberg data type names here, maybe that's more consistent with other type we put here, like `long`

> | *required*  | `return-type`              | `string` (Iceberg data type)                           | Return type. Example: `"string"`, `"struct<...>"`.                                            |

Reviewer 4: If this is an Iceberg data type, then it can't be a `string` because complex Iceberg types are represented as objects:

```json
{
  "type": "list",
  "element-id": 3,
  "element-required": true,
  "element": "string"
}
```

Iceberg does not parse these types other than to/from the JSON representation and that representation requires field IDs. If we want to omit the field IDs then that's fine but I think we need to be more specific. We don't want to need to parse complex types as strings -- we would need to define the string format and that would duplicate the JSON representation.

Author: Made the change accordingly. Added the requirement for returned values.

| *required*  | `versions`                 | `list<overload-version>`                               | [Versioned implementations](#overload-version) of this overload.                              |

> | *required*  | `current-overload-version` | `long`                                                 | Monotonically increasing identifier of the current overload version.                          |

Reviewer 4: Why use a `long`? Do we expect more than 2 billion versions of an overload?

| *optional*  | `function-type`            | `string` (either `"udf"` or `"udtf"`, default `"udf"`) | If `"udtf"`, `return-type` must be a `struct` describing the output schema.                   |
| *optional*  | `doc`                      | `string`                                               | Documentation string.                                                                         |

> ### Parameter

Reviewer 4: Do we want to mention that the parameters do not have default values?

Author: Added a note to call it out that parameters have no default value. Thanks for the suggestions. We may think about support it later. Snowflake[1] and Postgres support it. Spark[2] plans to support it. 

1. “Each optional argument has a default value that is used when the argument is omitted.” –[ Snowflake Docs](https://docs.snowflake.com/en/developer-guide/udf/udf-calling-sql) ·[ Snowflake Blog](https://medium.com/snowflake/snowflake-supports-named-and-optional-arguments-389d2500726f)
2. [UDFs with default values for parameters as a future improvement](https://www.databricks.com/blog/2021/10/20/introducing-sql-user-defined-functions.html)

Author: > define the cases where you must use the same set of parameters. 

This was done when we uses the type tuples as definition id. It guarantees that functions with the same type tuples cannot co-exist, even their return types are different.

| Requirement | Field  | Type                         | Description              |
|-------------|--------|------------------------------|--------------------------|
| *required*  | `name` | `string`                     | Parameter name.          |
| *required*  | `type` | `string` (Iceberg data type) | Parameter data type.     |

> | *optional*  | `doc`  | `string`                     | Parameter documentation. |

Reviewer 4: Is there a way to specify that null values are not allowed? In Iceberg nested types, each field/element/value is required or optional.

Author: Added one rule for inputs to clarify it.

Notes:

> 1. The `name` and `type` of a `parameter` are immutable. To change them, a new overload must be created. Only the optional documentation field (`doc`) can be updated in-place.

Author: > not sure if I understand how the parameter renaming cause problem for rollback.

Taking the example I put above. If we rename it to `y` at some point, then rollback to v1 or v2 will cause inconsistency between representation `x+1` and parameter name `y`.

> Is it ok to add an overload with only parameter name change, while params type and order are the same? How would client/engine resolve to the correct overload?

It shouldn't be allowed, as the signatures are the same.

> 1. The `name` and `type` of a `parameter` are immutable. To change them, a new overload must be created. Only the optional documentation field (`doc`) can be updated in-place.

Reviewer: got it. basically, parameter renaming is not allowed. hence, we require the `name` and `type` are immutable.

> 1. The `name` and `type` of a `parameter` are immutable. To change them, a new overload must be created. Only the optional documentation field (`doc`) can be updated in-place.

Author: Per last discussion(https://lists.apache.org/thread/t30hfxydwd8qkfzk9mtscc2xpg3kf621), we keep parameters only at the definition layer.

> 2. The `return-type` is immutable. To change it, users must create a new overload and deprecate or remove the old one.

Reviewer 4: What if the return type generated by one engine is different? For instance, what if I have an `index_of` function that looks for a value in a list. In one engine, lists are limited to < Integer.MAX_VALUE and in another they are not so the engines produce `int` and `long` return values. I think it would be okay to promote a definition that produces `int` to `long`. I'm also not sure how you would do it without replacing the value. Otherwise you'd have multiple definitions with the same parameters but different return types.

I think we want to avoid allowing different overloads/definitions with the same parameters right?

### Overload-Version
Each overload can evolve over time by introducing new versions. An overload version represents a specific implementation
of the overload at a given point in time.

| Requirement | Field name            | Type                                                                                 | Description                                                                             |
|-------------|-----------------------|--------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|

> | *required*  | `overload-version-id` | `long`                                                                               | Monotonically increasing identifier of the overload version.                            |

Reviewer 4: How about `version-id` since we don't need more than one version concept in this spec?

| *required*  | `representations`     | `list<representation>`                                                               | [Dialect-specific implementations](#representation).                                    |
| *optional*  | `deterministic`       | `boolean` (default `false`)                                                          | Whether the function is deterministic.                                                  |

> | *optional*  | `null-handling`       | `string` (either `"returns_null"` or `"called_on_null"`, default `"called_on_null"`) | Hint describing how the function behaves with NULL input values. See below for details. |

Reviewer 4: Is this a hint? I think if it is present then it needs to correctly describe behavior.

This also appears to be confusing to me. From a direct reading, I would expect `returns_null` to indicate that the function may return null for any input, not that for a particular input it MUST return null.

Author: Changed to `called_on_null_input` and `returns_null_on_null_input`. These are common terms used across engines, like Trinio, Microsoft SQL Server, and Snowflake. Keeping them consistent would be preferable. 

> Is this a hint?

It should not be a hint for engines supporting `null handling` flag, but it's practically a hint for engines not supporting it. For example, Spark will always be with `called_on_null_input` even the UDF is marked as `returns_null_on_null_input`, of course, the UDF authors can handle NULL explicitly inside function bodies. I have reword a bit to remove the `hint` here.

| *required*  | `dialect`  | `string` | SQL dialect identifier (e.g., `"spark"`, `"trino"`). |
| *required*  | `body`     | `string` | SQL expression text.                                 |

Note: The `body` must be valid SQL in the specified dialect; validation is the responsibility of the consuming engine.

### Definition log
| Requirement | Field name          | Type                                                     | Description                                                     |
|-------------|---------------------|----------------------------------------------------------|-----------------------------------------------------------------|
| *required*  | `timestamp-ms`      | `long` (epoch millis)                                    | When the definition snapshot was created or updated.            |
| *required*  | `overload-versions` | `list<{ overload-id: long, overload-version-id: long }>` | Mapping of each overload to its selected version at this point. |

## Function Resolution in Engines
Resolution rule is decided by engines, but engines SHOULD:
1. Prefer exact signature matches over casting or subtyping.

> 2. Allow numeric widening (e.g., `INT` --> `BIGINT/FLOAT`) by default.

Reviewer 4: When needed and where it does not conflict with the rule above.

---
title: "SQL UDF Spec"
---
<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

# Iceberg SQL UDF Spec

## Background and Motivation

A SQL user-defined function (UDF or UDTF) is a callable routine that accepts input parameters, executes a function body.
Depending on the function type, the result can be:

> - **UDFs** – return a single value, which may be a primitive type (e.g., `int`, `string`) or a non-primitive type (e.g., `struct`, `list`).

Author: Debated this with myself a bit. Should we still call it Scalar UDFs if it can return `struct`, `list`, etc.? My thoughts is to not mentioned `Scalar` to avoid confusion, I can remove them in the other places.

> - **UDFs** – return a single value, which may be a primitive type (e.g., `int`, `string`) or a non-primitive type (e.g., `struct`, `list`).

Author: Thinking a bit more, it should be OK to use "scalar" to refer one value as a counterpart of multiple rows. Made the change, thanks for the suggestion.

> - **Table functions (UDTFs)** – return a table, i.e., a table with zero or more rows and columns.

Reviewer 5: We might want to define the result zero or more rows with a uniform schema.

Many compute engines (e.g., Spark, Trino) already support UDFs, but in different and incompatible ways. Without a common
standard, UDFs cannot be reliably shared across engines or reused in multi-engine environments.

This specification introduces a standardized metadata format for UDFs in Iceberg. 

## Goals

* Define a portable metadata format for both scalar and table SQL UDFs. The metadata is self-contained and can be moved across catalogs. 
* Support function evolution through versioning and rollback.
* Provide consistent semantics for representing UDFs across engines.

## Overview

UDF metadata follows the same design principles as Iceberg table and view metadata: each function is represented by a
**self-contained metadata file**. Metadata captures definitions, parameters, return types, documentation, security,
properties, and engine-specific representations.

> * Any modification (new overload, updated representation, changed properties, etc.) creates a new metadata file. The old file is replaced atomically.

Author: That's a good point, let me reword it.

> * Any modification (new overload, updated representation, changed properties, etc.) creates a new metadata file. The old file is replaced atomically.

Author: Fixed

* Each metadata file includes recent definition versions, enabling rollbacks without external state.

## Specification

### UDF Metadata
The UDF metadata file has the following fields:

> | Requirement | Field name                   | Description                                                      |

Reviewer 5: I feel like we are missing the type information for each of the defined fields.  Are they a primitive (like a string/Boolean) or a structured/nested type defined below.

Author: Added type for each field, please take a look.

Notes:
1. The `name` and `type` of `parameters` are immutable. To change them, a new overload must be created. Only the optional documentation field (`doc`) can be updated in-place.
2. The `return type` is immutable. To change it, users must create a new overload and deprecate or remove the old one.

### Overload-Version
Each overload can evolve over time by introducing new versions. An overload version represents a specific implementation
of the overload at a given point in time.

| Requirement | Field name            | Description                                                                                  |
|-------------|-----------------------|----------------------------------------------------------------------------------------------|
| *required*  | `overload-version-id` | Monotonically increasing identifier (long) for this overload’s version history. Example: `1` |
| *required*  | `representations`     | Dialect-specific implementations of this overload.                                           |

> | *optional*  | `deterministic`       | Whether the function is deterministic. Defaults to `false`.                                  |

Author: Added it to the spec. Please take a look!

                     "body": "x + 2"
                  }
               ],
               "timestamp-ms": 1734507000123
            },
            {
               "overload-version-id": 2,
               "deterministic": true,
               "representations": [
                  {
                     "dialect": "trino",
                     "body": "x + 1"
                  }
               ],
               "timestamp-ms": 1735507000124
            }
         ],
         "current-overload-version": 2
      },
      {
         "overload-uuid": "7c9f93b1-28b4-4ef5-90f5-70c73cda2222",
         "parameters": [
            { "name": "x", "type": "float", "doc": "Input float" }
         ],
         "return-type": "float",
         "doc": "Add one to the input float",
         "versions": [
            {
               "overload-version-id": 1,
               "deterministic": true,
               "representations": [
                  {
                     "dialect": "trino",
                     "body": "x + 1.0"
                  }
               ],
               "timestamp-ms": 1734507001123
            }
         ],
         "current-overload-version": 1
      }
   ],

>    "definition-versions": [

Author: Per community sync, changed it to `definition-log`, which doesn't keep lineage of the global version change. `timestamp` is still used to indicate when the change happens.

| *required*  | `function-uuid`  | `string`               | A UUID that identifies the function, generated once at creation. |
| *required*  | `format-version` | `int`                  | Metadata format version (must be `1`).                           |
| *required*  | `definitions`    | `list<definition>`     | List of function [definition](#definition) entities.             |
| *required*  | `definition-log` | `list<definition-log>` | History of [definitions snapshots](#definitions-log).            |
| *optional*  | `location`       | `string`               | Storage location of metadata files.                              |
| *optional*  | `properties`     | `map`                  | A string to string map of properties.                            |
| *optional*  | `secure`         | `boolean`              | Whether it is a secure function. Default: `false`.               |
| *optional*  | `doc`            | `string`               | Documentation string.                                            |

Notes:
1. When `secure` is `true`:
   - Engines **SHOULD NOT** expose the function definition through inspection (e.g., `SHOW FUNCTIONS`).

>    - Engines **SHOULD** ensure that execution does not leak sensitive information through error messages, logs, or query plans.

Author: 1. Good point. I think the intent is that if an engine declares support for `secure=true`, it must enforce these requirements; otherwise it should reject the function as unsupported. That way, users can rely on consistent behavior instead of silent non-compliance.
2. Yes, I agree we should mention that secure UDFs must be executed atomically and not be reordered or optimized away in predicate pushdowns. I’ll add a note.

Notes:
1. `sig1-<base32(SHA-256(signature))[:26]>` is a fixed-length hash used when the canonical signature is too long.
It’s generated by taking the SHA-256 of the normalized signature, encoding it in Base32, keeping the first 26 characters,
and prefixing with `sig1-`. This yields a 31-character deterministic ID, easy to verify and future-proof via the `sig1-`
version prefix.

### Parameter
| Requirement | Field  | Type                                                                                                                                                           | Description              |
|-------------|--------|----------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------|

> | *required*  | `name` | `string`                                                                                                                                                       | Parameter name.          |

Reviewer 6: Are the names need to be considered as case-sensitive or in sensitive ?

| *optional*  | `parameters` | `list<parameter>` | Ordered list of [function parameters](#parameter). Overrides canonical names in definition. |
| *required*  | `body`       | `string`          | SQL expression text.                                                                        |

Note: The `body` must be valid SQL in the specified dialect; validation is the responsibility of the consuming engine.

### Definition log
| Requirement | Field name            | Type                                               | Description                                                      |
|-------------|-----------------------|----------------------------------------------------|------------------------------------------------------------------|
| *required*  | `timestamp-ms`        | `long` (epoch millis)                              | When the definition snapshot was created or updated.             |
| *required*  | `definition-versions` | `list<{ definition-id: string, version-id: int }>` | Mapping of each definition to its selected version at this time. |

## Function Resolution in Engines

> Resolution rule is decided by engines, but engines SHOULD:

Author: Right, and the spec can only minimize that.

| *required*  | `format-version` | `int`                  | Metadata format version (must be `1`).                           |
| *required*  | `definitions`    | `list<definition>`     | List of function [definition](#definition) entities.             |
| *required*  | `definition-log` | `list<definition-log>` | History of [definitions snapshots](#definitions-log).            |
| *optional*  | `location`       | `string`               | Storage location of metadata files.                              |
| *optional*  | `properties`     | `map`                  | A string to string map of properties.                            |
| *optional*  | `secure`         | `boolean`              | Whether it is a secure function. Default: `false`.               |
| *optional*  | `doc`            | `string`               | Documentation string.                                            |

Notes:
1. When `secure` is `true`:
   - Engines MUST NOT expose the function definition or its body through any form of metadata inspection (e.g., `SHOW FUNCTIONS`).
   - Engines MUST prevent leakage of sensitive information during execution via error messages, logs, query plans, or intermediate results.

>    - Engines MUST NOT perform predicate reordering, short-circuiting, or other optimizations that could change the order or scope of data access.

Author: Good point. The current rules imply that already. I can add these words if people strongly feel about it.

2. Parameter `name`s are immutable since named argument invocation is supported (e.g., `foo(a => 1, b => 2)`). Only `doc` can be updated in place. 
3. Variadic (vararg) parameters are not supported. Each definition must declare a fixed number of parameters.
4. Each parameter input MUST be assignable to its declared Iceberg type. For complex types, the value’s
   structure must match (correct field names, element/key/value types, and nesting). If a parameter—or any nested
   field/element—is marked required, engines MUST reject null at that position (including inside structs, lists, and maps).
5. The `return-type` is immutable. To change it, users must create a new definition and remove the old one.
6. The function MUST return a value assignable to the declared `return-type`, meaning the returned value’s type and
   structure must match the declared Iceberg type (including field names, element types, and nesting for complex types),
   and any field or element marked as required MUST NOT be null. Engines MUST reject results that violate these rules.

### Definition-Version

> Each definition can evolve over time by introducing new versions.  

Reviewer 7: Shall we put a cross-version consistency rule in this section:

- function-type MUST NOT change across versions of the same definition.
- parameter types and names, return-type  MUST remain identical across versions of a given definition. Changing them requires a new definition.

Author: That's a good point. Updated this section to the intent clearer and more explicit.

| *required*  | `parameters`         | `list<parameter>`                                                                                                                                              | Ordered list of [function parameters](#parameter). Invocation order **must** match this list.                                                                                                       |
| *required*  | `return-type`        | [JSON representation](https://iceberg.apache.org/spec/#appendix-c-json-serialization) of an Iceberg type (`string` for primitives, `object` for complex types) | Type of value returned                                                                                                                                                                              |
| *optional*  | `nullable-return`    | `boolean`                                                                                                                                                      | A hint to indicate whether the return value is nullable or not. Default: `true`.                                                                                                                    |

> | *required*  | `versions`           | `list<definition-version>`                                                                                                                                     | [Versioned implementations](#definition-version) of this definition.                                                                                                                                |

Author: Yes, each `definition-version` is immutable, we need to introduce a new version for changes.

| *required*  | `body`       | `string`          | SQL expression text.                                                                        |

Note: The `body` must be valid SQL in the specified dialect; validation is the responsibility of the consuming engine.

### Definition log
| Requirement | Field name            | Type                                               | Description                                                      |
|-------------|-----------------------|----------------------------------------------------|------------------------------------------------------------------|
| *required*  | `timestamp-ms`        | `long` (epoch millis)                              | When the definition snapshot was created or updated.             |
| *required*  | `definition-versions` | `list<{ definition-id: string, version-id: int }>` | Mapping of each definition to its selected version at this time. |

## Function Resolution in Engines
Resolution rule is decided by engines, but engines SHOULD:
1. Prefer exact signature matches over casting or subtyping.

> 2. Allow numeric widening (e.g., `INT` --> `BIGINT/FLOAT`) when needed and where it does not conflict with the rule above.

Reviewer 8: Maybe I miss something, not sure I see what situation would it conflict with rule above?

Also, is this tied to Iceberg schema evolution rules, or is it separate?

> 3. Require explicit casts for unsafe conversions (e.g., `STRING` --> `DATE`).

Author: Trying to understand this, are you saying engines couldn't break the tie when there are two overloads like this, and caller get a `int` input? We discussed this a bit previously.  
```
foo(bigint)
foo(double)
```

The general thought goes to let engine make the decision, as you can see in the beginning 
```
Resolution rule is decided by engines, but engines SHOULD:
```
These are NOT must-follow rules.

**self-contained metadata file**. Metadata captures definitions, parameters, return types, documentation, security,
properties, and engine-specific representations.

* Any modification (new definition, updated representation, changed properties, etc.) creates a new metadata file, and atomically swaps in the new file as the current metadata.
* Each metadata file includes recent definition versions, enabling rollbacks without external state.

## Specification

### UDF Metadata
The UDF metadata file has the following fields:

| Requirement | Field name        | Type                   | Description                                                           |
|-------------|-------------------|------------------------|-----------------------------------------------------------------------|
| *required*  | `function-uuid`   | `string`               | A UUID that identifies the function, generated once at creation.      |
| *required*  | `format-version`  | `int`                  | Metadata format version (must be `1`).                                |
| *required*  | `definitions`     | `list<definition>`     | List of function [definition](#definition) entities.                  |

> | *required*  | `definition-log`  | `list<definition-log>` | History of [definition snapshots](#definition-log).                   |

Reviewer 4: It isn't clear what a "definition snapshot" is without going to that section. How about "Definition version history" for the description instead?

| *optional*  | `location`        | `string`               | The function's base location; used to create metadata file locations. |
| *optional*  | `properties`      | `map<string,string>`   | A string-to-string map of properties.                                 |
| *optional*  | `secure`          | `boolean`              | Whether it is a secure function. Default: `false`.                    |
| *optional*  | `doc`             | `string`               | Documentation string.                                                 |

Notes:

> 1. Engines must prevent leakage of sensitive information when a function is marked as `secure` by setting it to `true`.

Reviewer 2: If i'm an engine reading this note or property, what should I do? What is "sensitive information? " what is "leakage"?

| *required*  | `parameters`         | `list<parameter>`                               | Ordered list of [function parameters](#parameter). Invocation order **must** match this list.                 |
| *required*  | `return-type`        | `string`                                        | Declared return type (see [Parameter Type](#parameter-type)).                                                 |
| *optional*  | `nullable-return`    | `boolean`                                       | A hint to indicate whether the return value is nullable or not. Default: `true`.                              |
| *required*  | `versions`           | `list<definition-version>`                      | [Versioned implementations](#definition-version) of this definition.                                          |
| *required*  | `current-version-id` | `int`                                           | Identifier of the current version for this definition.                                                        |

> | *optional*  | `function-type`      | `string` (`"udf"` or `"udtf"`, default `"udf"`) | If `"udtf"`, `return-type` must be an Iceberg type `struct` describing the output schema.                     |

Reviewer 2: Why is this at the definition level? Are we ok with some signatures being UDF and others being UDTF?

| *optional*  | `doc`                | `string`                                        | Documentation string.                                                                                         |

### Parameter
| Requirement | Field  | Type     | Description                                                  |
|-------------|--------|----------|--------------------------------------------------------------|

> | *required*  | `type` | `string` | Parameter data type (see [Parameter Type](#parameter-type)). |

Reviewer 9: Do we allow nullable parameter?

3. Each parameter input MUST be assignable to its declared Iceberg type. For complex types, the value’s
   structure must match (correct field names, element/key/value types, and nesting). If a parameter—or any nested
   field/element—is marked required, engines MUST reject null at that position (including inside structs, lists, and maps).
4. The `return-type` is immutable across versions. To change it, users must create a new definition and remove the old one.
5. The function MUST return a value assignable to the declared `return-type`, meaning the returned value’s type and
   structure must match the declared Iceberg type (including field names, element types, and nesting for complex types),
   and any field or element marked as required MUST NOT be null. Engines MUST reject results that violate these rules.

#### Parameter-Type
Primitive types are encoded using the [Iceberg Type JSON Representation](https://iceberg.apache.org/spec/#appendix-c-json-serialization),
for example `"int"`, `"string"`.

Three composite types are supported. 
* `list<T>`: Ordered, homogeneous collection of elements of type `T` (any Iceberg type); example `list<int>`.

> * `map<K,V>`: Key–value collection; `K` must be a primitive, comparable type (`string` recommended); example `map<string,int>`.

Reviewer 4: We should be clear that there is not a way to leave the type parameters (like `T`) unbound. These must be concrete types.

|-------------|-------------------|---------------------------------------------------------------------------------------------------------|----------------------------------------------------------------|
| *required*  | `version-id`      | `int`                                                                                                   | Monotonically increasing identifier of the definition version. |
| *required*  | `representations` | `list<representation>`                                                                                  | [Dialect-specific implementations](#representation).           |
| *optional*  | `deterministic`   | `boolean` (default `false`)                                                                             | Whether the function is deterministic.                         |
| *optional*  | `on-null-input`   | `string` (`"returns_null_on_null_input"` or `"called_on_null_input"`, default `"called_on_null_input"`) | Defines how the UDF behaves when any input parameter is NULL.  |

> | *required*  | `timestamp-ms`    | `long` (epoch millis)                                                                                   | Creation timestamp of this version.                            |

Reviewer 4: What epoch? I think you can borrow some language from the table spec.

1. If set to `returns_null_on_null_input`, the function always returns `NULL` if any input argument is `NULL`. This allows engines to apply predicate pushdown or skip function evaluation for rows with `NULL` inputs. For a function `f(x, y) = x + y`,
the engine can safely rewrite `WHERE f(a,b) > 0` as `WHERE a IS NOT NULL AND b IS NOT NULL AND f(a,b) > 0`.
2. If set to `called_on_null_input`, the function may handle `NULL`s internally (e.g., `COALESCE`, `NVL`, `IFNULL`), so the engine must execute the function even if some inputs are `NULL`.

### Representation
A representation encodes how the definition version is expressed in a specific SQL dialect.

| Requirement | Field name   | Type              | Description                                                                                 |
|-------------|--------------|-------------------|---------------------------------------------------------------------------------------------|
| *required*  | `type`       | `string`          | Must be `"sql"`                                                                             |

> | *required*  | `dialect`    | `string`          | SQL dialect identifier (e.g., `"spark"`, `"trino"`).                                        |

Reviewer 4: The view spec says that only one representation per dialect is allowed:

> A view version can have multiple SQL representations of different dialects, but only one SQL representation per dialect.

Do we want to have the same restriction?

Author: Makes sense. Let me add it.

> | *required*  | `body`       | `string`          | SQL expression text.                                                                        |

Author: When we try to expand the spec to other representations, like Python UDF, the term `body` would be more generic.

Note: The `body` must be valid SQL in the specified dialect; validation is the responsibility of the consuming engine.

> ### Definition-Log

Reviewer 4: I don't think that it is helpful to have `-` in section headings like this. Either match the property name exactly (i.e. `definition-log`) or omit the `-` to make it human-readable ("Definition log entries").

| Requirement | Field name            | Type                                               | Description                                                      |
|-------------|-----------------------|----------------------------------------------------|------------------------------------------------------------------|
| *required*  | `timestamp-ms`        | `long` (epoch millis)                              | When the definition snapshot was created or updated.             |
| *required*  | `definition-versions` | `list<{ definition-id: string, version-id: int }>` | Mapping of each definition to its selected version at this time. |

## Function Call Convention and Resolution in Engines

> Resolution rule is decided by engines, but engines SHOULD:

Reviewer 4: I think it would be more clear to say this:

> Selecting the definition of a function to use is delegated to engines, which may apply their own casting rules. However, engines should:
> 1. Prefer exact parameter matches over safe (widening) or unsafe casts
> 2. Safely widen types as needed to avoid failing to find a matching definition
> 3. Require explicit casts for unsafe or non-obvious conversions
> 4. Use definitions with the same number of arguments as the input
> 5. Pass positional arguments in the same position as the input
> 6. Use definitions with the same set of field names as named input arguments

As for the last point of specifically not mixing positional and named arguments, I think that points 5 and 6 cover it. Don't reorder positional arguments and match the whole set of names. Also, implementers may ignore the "don't mix positional and named matching" but clearly stating how to match positional and named at least gives us some insurance that behavior won't be wacky if people do it anyway.

* Map: `map<key-type,value-type>` (e.g., `map<string,int>`)
* Struct: `struct<name1:type1,name2:type2,...>` with field names and types (e.g., `struct<id:int,name:string>`)

Examples of complete definition-id signatures:
* `int` – single int parameter
* `int,string` – two parameters: int and string
* `int,list<int>,struct<id:int,name:string>` – three parameters: an int, a list and a struct

### Definition Version

Each definition can evolve over time by introducing new versions.  
A `definition version` represents a specific implementation of that definition at a given point in time. A definition version can have multiple SQL representations of different dialects, but only one SQL representation per dialect.

| Requirement | Field name        | Type                                                     | Description                                                    |
|-------------|-------------------|----------------------------------------------------------|----------------------------------------------------------------|
| *required*  | `version-id`      | `int`                                                    | Monotonically increasing identifier of the definition version. |

> | *required*  | `representations` | `list<representation>`                                   | [Dialect-specific implementations](#representation).           |

Reviewer 4: The description here should not be "dialect-specific" because not all representations have a "dialect" -- in my opinion, SQL is the only one that would need a dialect.

| *required*  | `dialect`  | `string` | SQL dialect identifier (e.g., `"spark"`, `"trino"`). |
| *required*  | `sql`      | `string` | SQL expression text.                                 |

Notes:
1. The `sql` must reference parameters using the names declared in the definition's `parameters` field.

### Definition Log
| Requirement | Field name            | Type                                                | Description                                                      |
|-------------|-----------------------|-----------------------------------------------------|------------------------------------------------------------------|
| *required*  | `timestamp-ms`        | `long` (unix epoch millis)                          | When the definition snapshot was created or updated.             |
| *required*  | `definition-versions` | `list<struct<definition-id:string,version-id:int>>` | Mapping of each definition to its selected version at this time. |

## Function Call Convention and Resolution in Engines
Selecting the definition of a function to use is delegated to engines, which may apply their own casting rules. However, engines should:

> 1. Prefer exact parameter matches over safe (widening) or unsafe casts.

Reviewer 4: Minor: Point 3 says engines should require explicit casts for unsafe conversions, which might make this less clear because an explicit cast changes the input being matched. For example, when matching `foo(cast(string_col as int))`, the cast makes `int` an exact match. I think this guidance still applies when the engine doesn't strictly follow point 3, though. I'm on the fence about whether to only mention "safe (widening) casts" or to use the current version.

   the engine can safely rewrite `WHERE f(a,b) > 0` as `WHERE a IS NOT NULL AND b IS NOT NULL AND f(a,b) > 0`.
2. If set to `call`, the function may handle `NULL`s internally (e.g., `COALESCE`, `NVL`, `IFNULL`), so the engine must
   execute the function even if some inputs are `NULL`.

### Representation
Each representation is an object with at least one common field, `type`, that is one of the following:
* `sql`: a SQL expression that defines the function body

Representations further define metadata for each type.

#### SQL Representation

A definition version can have multiple SQL representations of different dialects, but only one SQL representation per dialect.
The SQL representation stores the function body as a SQL expression, with metadata such as the SQL dialect.

| Requirement | Field name | Type     | Description                                          |
|-------------|------------|----------|------------------------------------------------------|
| *required*  | `type`     | `string` | Must be `"sql"`                                      |
| *required*  | `dialect`  | `string` | SQL dialect identifier (e.g., `"spark"`, `"trino"`). |

> | *required*  | `sql`      | `string` | SQL expression text.                                 |

Author: Discussed offline with @singhpk234. We chose to follow the view spec and intentionally avoid specifying this further. There is a fine balance between providing enough clarity in the specification and over specifying behavior that is better left to implementations.

## Specification

### UDF Metadata
The UDF metadata file has the following fields:

| Requirement | Field name        | Type                   | Description                                                                                                     |
|-------------|-------------------|------------------------|-----------------------------------------------------------------------------------------------------------------|
| *required*  | `function-uuid`   | `string`               | A UUID that identifies the function, generated once at creation.                                                |
| *required*  | `format-version`  | `int`                  | Metadata format version (must be `1`).                                                                          |
| *required*  | `definitions`     | `list<definition>`     | List of function [definition](#definition) entities.                                                            |
| *required*  | `definition-log`  | `list<definition-log>` | History of [definition snapshots](#definition-log).                                                             |

> | *required*  | `parameter-names` | `list<parameter-name>` | Global ordered parameter names shared across all overloads. Overloads must use a prefix of this list, in order. |

Reviewer 10: thanks.

| *optional*  | `location`        | `string`               | Storage location of metadata files.                                                                             |
| *optional*  | `properties`      | `map<string,string>`   | A string-to-string map of properties.                                                                           |
| *optional*  | `secure`          | `boolean`              | Whether it is a secure function. Default: `false`.                                                              |
| *optional*  | `doc`             | `string`               | Documentation string.                                                                                           |

> ### Parameter-Name

Reviewer: the notes for `secure` and `properties` should belong to the previous section. should the `Parameter-Name` subsection be moved after the `Notes:`? the notes for the `parameter-names` can be moved to that section too.

Reviewer 4: Shouldn't this be `parameter` that has a `name` and a `doc`? A parameter name that contains a name is redundant.

| Requirement | Field  | Type     | Description              |
|-------------|--------|----------|--------------------------|
| *required*  | `name` | `string` | Parameter name.          |
| *optional*  | `doc`  | `string` | Parameter documentation. |

Notes:
1. When `secure` is `true`:
   - Engines MUST NOT expose the function definition or its body through any form of metadata inspection (e.g., `SHOW FUNCTIONS`).
   - Engines MUST prevent leakage of sensitive information during execution via error messages, logs, query plans, or intermediate results.
   - Engines MUST NOT perform predicate reordering, short-circuiting, or other optimizations that could change the order or scope of data access.

> 2. Entries in `properties` are treated as hints, not strict rules. Engines MAY choose to honor them or ignore them.

Author: Removed it here. Let me know if a global explanation is needed.

   invocation consistent across definitions:
   - Each overload uses the first N entries of this list for its arity, in order.
   - Names and their relative ordering are immutable. Only appending new names is allowed.
   - Only the `doc` field may be updated in place.
   - Overloads that differ only by parameter types are allowed (e.g., `foo(int)`, `foo(float)`, `foo(string, string)`), and
     they all derive their parameter names positionally from the same `parameter-names` prefix.

### Definition

Each `definition` represents one function signature (e.g., `add_one(int)` vs `add_one(float)`).

| Requirement | Field name           | Type                                                                                                                                                                                                                                                        | Description                                                                                                                                                                                         |

> |-------------|----------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

Reviewer 4: Minor: In the table spec, we avoid table wrapping issues by omitting whitespace from the last field. That would help readability here.

> | *required*  | `definition-id`      | `string`                                                                                                                                                                                                                                                    | An identifier derived from canonical parameter-type tuple (lowercase, no spaces; e.g., `"(int,int,string)"`). If longer than 128 chars, use hashed form `"sig1-<base32(SHA-256(signature))[:26]>"`. |

Author: Lookup misses should happen if the implementation follows the spec and normalizes(hashes) the signature before lookup. As long as both insertion and lookup go through the same hashing step, the hash signature should remain consistent.

That said, I’m definitely open to suggestions for a shorter or cleaner representation if folks have ideas, happy to iterate here.

| *required*  | `parameters`         | `list<parameter>`                                                                                                                                                                                                                                           | Ordered list of [function parameters](#parameter). Invocation order **must** match this list.                                                                                                       |

> | *required*  | `return-type`        | [Iceberg Type JSON Representation](https://iceberg.apache.org/spec/#appendix-c-json-serialization), primitive types are encoded as JSON strings (e.g., `"int"`, `"string"`), and complex types are encoded as JSON objects (e.g., `{"type": "struct", ...}` | Type of value returned                                                                                                                                                                              |

Reviewer 4: Tables are really hard to work with when they have embedded URLs like this. I think a best practice is to use a named URL:

```
[text][link-name]
...
[link-name]: https://...
```

I also highly recommend keeping table content short and minimal. If you need more clarification for how a field is used or interpreted, the additional information should go in a paragraph or number list of notes after the table.

Author: I will reword a bit.

> | *optional*  | `nullable-return`    | `boolean`                                                                                                                                                                                                                                                   | A hint to indicate whether the return value is nullable or not. Default: `true`.                                                                                                                    |

Reviewer 4: I agree. Nullable by default.

| *required*  | `versions`           | `list<definition-version>`                                                                                                                                                                                                                                  | [Versioned implementations](#definition-version) of this definition.                                                                                                                                |
| *required*  | `current-version-id` | `int`                                                                                                                                                                                                                                                       | Identifier of the current version for this definition.                                                                                                                                              |

> | *optional*  | `function-type`      | `string` (`"udf"` or `"udtf"`, default `"udf"`)                                                                                                                                                                                                             | If `"udtf"`, `return-type` must be an Iceberg type `struct` describing the output schema.                                                                                                           |

Author: I can make this field required if the consensus is that it is better, but I do not think the current behavior is confusing.

To clarify the question about optionality: even if the field is optional in the spec, clients are still expected to follow the spec correctly. In this case, when a user defines a UDTF, the client must explicitly set this field to udtf. Omitting it would be a spec violation.

This is consistent with how we handle other optional fields that carry semantic meaning. Optional does not imply free to omit arbitrarily. It means the field may be absent only when its default semantics are correct. For UDTFs, the default semantics are not correct, so compliant clients must set it.

Given that this is entirely under client control and engines can reliably infer whether a function is scalar or tabular at creation time, I do not see a practical risk of ambiguity here.

> | *optional*  | `function-type`      | `string` (`"udf"` or `"udtf"`, default `"udf"`)                                                                                                                                                                                                             | If `"udtf"`, `return-type` must be an Iceberg type `struct` describing the output schema.                                                                                                           |

Reviewer 7: I think it’s valid for function-type to be optional overall (default udf), but writers MUST set it to udtf when defining a UDTF.

> | *optional*  | `function-type`      | `string` (`"udf"` or `"udtf"`, default `"udf"`)                                                                                                                                                                                                             | If `"udtf"`, `return-type` must be an Iceberg type `struct` describing the output schema.                                                                                                           |

Reviewer 10: I think it all boils down to what does `optional` mean here? 

You guys are saying optional means all implementation writers should write in one scenario (UDTF) and skip in other scenario (UDF). IMO, optional means any writer can choose to write or omit the field irrespective of scenarios. So, in UDTF case itself writers can write or omit if it is optional field. This was my understanding all along. 

We probably need to update the [writer requirement ](https://iceberg.apache.org/spec/#writer-requirements) in the spec if my understanding is wrong. Because for some people (like me or maybe you) it is not clear.

> | *optional*  | `function-type`      | `string` (`"udf"` or `"udtf"`, default `"udf"`)                                                                                                                                                                                                             | If `"udtf"`, `return-type` must be an Iceberg type `struct` describing the output schema.                                                                                                           |

Reviewer: @ajantha-bhat I don't see conflict with the linked writer requirement. This field can be omitted. In this case, it is just implied that the function type is udf. It is a like defining an optional field in SQL with default value. I am not seeing any problem here.

> | *optional*  | `function-type`      | `string` (`"udf"` or `"udtf"`, default `"udf"`)                                                                                                                                                                                                             | If `"udtf"`, `return-type` must be an Iceberg type `struct` describing the output schema.                                                                                                           |

Reviewer 10: > one example would be [snapshot ref](https://iceberg.apache.org/spec/#snapshot-references). max-ref-age-ms is optional in the snapshot ref object. it is default to table property history.expire.max-ref-age-ms. I see it the same thing here.

We aren't forcing user that `max-ref-age-ms` must be always set in some scenario. Here it is different, we are forcing user that UDTF must be set in UDTF scenario. 

So, If somethings are MUST be set, make it a required field. Thats my personal preference of working with IETF RFCs in the past for optional fields. I derived the same meaning from writer requirements in the [spec](https://iceberg.apache.org/spec/#writer-requirements).  

But again if the most of the members have aligned towards forcing user to set something for optional fields, I am ok with it. You guys can go ahead on this change. Community consensus matters than individual opinion or preference.

> | *optional*  | `function-type`      | `string` (`"udf"` or `"udtf"`, default `"udf"`)                                                                                                                                                                                                             | If `"udtf"`, `return-type` must be an Iceberg type `struct` describing the output schema.                                                                                                           |

Author: Changed it to `required`

---
title: "SQL UDF Spec"
---
<!--
 - Licensed to the Apache Software Foundation (ASF) under one or more
 - contributor license agreements.  See the NOTICE file distributed with
 - this work for additional information regarding copyright ownership.
 - The ASF licenses this file to You under the Apache License, Version 2.0
 - (the "License"); you may not use this file except in compliance with
 - the License.  You may obtain a copy of the License at
 -
 -   http://www.apache.org/licenses/LICENSE-2.0
 -
 - Unless required by applicable law or agreed to in writing, software
 - distributed under the License is distributed on an "AS IS" BASIS,
 - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 - See the License for the specific language governing permissions and
 - limitations under the License.
 -->

# Iceberg UDF Spec

## Background and Motivation

A SQL user-defined function (UDF or UDTF) is a callable routine that accepts input parameters and executes a function body.
Depending on the function type, the result can be:

> - **Scalar function (UDF)** – returns a scalar value, which may be a primitive type (e.g., `int`, `string`) or a non-primitive type (e.g., `struct`, `list`).

Reviewer 4: I think it would be more clear to say "returns a single value". The term "scalar" gets a bit weird because in other contexts it means "not a vector" or "not a matrix" but we allow structs and lists to be returned. I think it's okay to call the function a "scalar function", but I'd like to be clear that the "scalar" part just means a single value is returned, even if that value may have nested fields.

| *required*  | `parameter-names` | `list<parameter-name>` | Global ordered parameter names shared across all overloads. Overloads must use a prefix of this list, in order. |
| *optional*  | `location`        | `string`               | The function's base location; used to create metadata file locations.                                           |
| *optional*  | `properties`      | `map<string,string>`   | A string-to-string map of properties.                                                                           |
| *optional*  | `secure`          | `boolean`              | Whether it is a secure function. Default: `false`.                                                              |
| *optional*  | `doc`             | `string`               | Documentation string.                                                                                           |

Notes:
1. When `secure` is `true`:
   - Engines MUST NOT expose the function definition or its body through any form of metadata inspection (e.g., `SHOW FUNCTIONS`).
   - Engines MUST prevent leakage of sensitive information during execution via error messages, logs, query plans, or intermediate results.

>    - Engines MUST NOT perform predicate reordering, short-circuiting, or other optimizations that could change the order or scope of data access.

Reviewer 4: I'm not sure that these engine requirements should be in the spec. The first one is easy to poke holes in -- if an engine knows who can see and modify a function definition then it should not apply.

The third one also seems overly broad. I would prefer to have a higher-level statement like "engines should not apply optimizations that may leak information". It seems to me that engines could easily make different choices about what optimizations are applied and could be clever.

What about something like "Engines should prevent leakage of sensitive information when a function is marked as `secure`"? The way to do that seems fundamentally like an engine choice to me.

Author: Make sense to have more general description. I think we want this to be a rule other than a hint. Made the change and reword a bit, like `Engines must prevent leakage of sensitive information when`.
